### PR TITLE
Swap-rewards kafka consumer

### PR DESCRIPTION
**Pull request by v-kat**
### Summary

Kafka consumer for swap-rewards project.
Information still needs to be fleshed-out.

Also thought that https://github.com/brave-intl/bat-go/blob/master/utils/kafka/dialer.go#L49 probably wasn't the default we want for everything.... I used a sentinel wallet that we could insert in a migration (I can throw that in 46 or we could do another approach).

Sorry for merging this into another branch.

### Type of change ( select one )

- [X] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### CODE REVIEW

**Comment by v-kat** [comment]
@nvonpentz are you going to use an idempotent producer too? I'm not sure if the Go library we're using supports this but I also don't know where you're producing from https://github.com/segmentio/kafka-go/issues/797

File: docker-compose.yml [hunk 1]
```diff
@@ -43,6 +44,7 @@ services:
       - KAFKA_SSL_KEY_LOCATION=/etc/kafka/secrets/consumer.client.key
       - KAFKA_REQUIRED_ACKS=1
       - KAFKA_CONSUMER_GROUP_PROMOTIONS=grant-bat-promotions-local
+      - KAFKA_CONSUMER_GROUP_REWARDs=grant-bat-rewards-local
```
Comment (husobee) [inline docker-compose.yml:47]: KAFKA_CONSUMER_GROUP_REWARDS as the env variable instead?

File: migrations/0046_swap_rewards_kafka.up.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+--- swap promotions grant reward fields for kafka worker
+alter table claims add column transaction_key uuid;
+create unique index unique_claim_hack_idx on claims(transaction_key) where transaction_key is not NULL;
```
Comment (husobee) [inline migrations/0046_swap_rewards_kafka.up.sql:3]: naming, maybe something like `unique_claim_transaction_key`

File: promotion/datastore.go [hunk 3]
```diff
@@ -1804,6 +1806,46 @@ func (pg *Postgres) RunNextDrainRetryJob(ctx context.Context, worker DrainRetryW
 	}
 }
 
+// RunNextFetchRewardGrantsJob - runs fetch reward grant jobs
+func (pg *Postgres) RunNextFetchRewardGrantsJob(ctx context.Context, worker SwapRewardsWorker) error {
+
+	service, _ := InitService(ctx, pg, nil, nil)
```
Comment (husobee) [inline promotion/datastore.go:1812]: Not sure we want to InitService again?  I think InitService is really only meant to be run once on service startup.

File: promotion/datastore.go [hunk 4]
```diff
@@ -1804,6 +1806,46 @@ func (pg *Postgres) RunNextDrainRetryJob(ctx context.Context, worker DrainRetryW
...
+	}
+
+	cohort := "control"
+	//maybe this call should be changed to just rely on the right status codes and only do 1 post request
+	_, err = service.GetOrCreateIssuer(ctx, grant.PromotionID, cohort)
+	if err != nil {
+		return err
+	}
+
+	//TODO below somewhere else and not have magic strings
```
Comment (husobee) [inline promotion/datastore.go:1826]: Agree, lets put this elsewhere.

File: promotion/service.go [hunk 5]
```diff
@@ -35,10 +35,13 @@ const localEnv = "local"
 var (
 	// toggle for drain retry job
 	enableDrainRetryJob = isDrainRetryJobEnabled()
+	// toggle for grant rewards job
+	enableRewardGrantsJob = isGrantRewardsJobEnabled()
 	// toggle for gemini check status
 	enableGeminiCheckStatus = isRunNextGeminiCheckStatus()
 
 	suggestionTopic       = os.Getenv("ENV") + ".grant.suggestion"
+	rewardsTopic          = os.Getenv("ENV") + ".grant.rewards"
```
Comment (husobee) [inline promotion/service.go:44]: should we explicitly have this topic say swaprewards?

File: promotion/service.go [hunk 6]
```diff
@@ -174,6 +198,27 @@ func (service *Service) InitKafka(ctx context.Context) error {
 		adminAttestationTopic: adminAttestationEventSchema,
 	})
 
+	if err != nil {
+		return fmt.Errorf("failed to generate codecs kafka: %w", err)
+	}
+
+	// toggle for grant reward job
+	if enableRewardGrantsJob {
+		groupID := os.Getenv("KAFKA_CONSUMER_GROUP_REWARDs")
```
Comment (husobee) [inline promotion/service.go:207]: KAFKA_CONSUMER_GROUP_REWARDS instead?

File: promotion/datastore.go (hunk 3 above)
Comment (v-kat) [reply promotion/datastore.go:1812]: I wasn't sure of a better way to go from `pg` and `ctx` to `service` that was needed further down.

Comment (v-kat) [reply promotion/datastore.go]: https://github.com/brave-intl/bat-go/pull/1233/commits/dc36d86c2d9ba0e0190d72478e3b5af177bef05d

File: promotion/service.go [hunk 7]
```diff
@@ -35,10 +35,13 @@ const localEnv = "local"
 var (
 	// toggle for drain retry job
 	enableDrainRetryJob = isDrainRetryJobEnabled()
+	// toggle for grant rewards job
+	enableRewardGrantsJob = isGrantRewardsJobEnabled()
 	// toggle for gemini check status
 	enableGeminiCheckStatus = isRunNextGeminiCheckStatus()
 
 	suggestionTopic       = os.Getenv("ENV") + ".grant.suggestion"
+	rewardsTopic          = os.Getenv("ENV") + ".grant.swaprewards"
```
Comment (v-kat) [inline promotion/service.go:44]: @nvonpentz

**Comment by v-kat** [comment]
Got an approval from @husobee  in slack.

**Pull request merged**